main/modes/timer/modeTimer.c

```diff
@@ -0,0 +1,286 @@
+
+    /// @brief Current timer state
+    timerState_t timerState;
+
+    /// @brief If true, counts up (stopwatch), otherwise counts down (timer)
+    bool stopwatch;
+
+    /// @brief The amount of time being counted down from
+    int64_t countdownTime;
+
+    /// @brief Any time that was accumulated before the timer was resumed
+    int64_t accumulatedDuration;
+
+    /// @brief The actual time the timer was started
+    int64_t startTime;
+} timerMode_t;
+
+//==============================================================================
+// Function Prototypes
+//==============================================================================
+
+static void timerEnterMode(void);
+static void timerExitMode(void);
+static void timerMainLoop(int64_t elapsedUs);
+
+//==============================================================================
+// Strings
+//==============================================================================
+
+static const char timerName[]              = "Timer";
+static const char hourglassFrameFmt[]      = "hourglass_%02" PRIu8 ".wsg";
+static const char minutesSecondsFmt[]      = "%" PRIu8 ":%02" PRIu8 ".";
+static const char hoursMinutesSecondsFmt[] = "%" PRIu64 ":%02" PRIu8 ":%02" PRIu8 ".";
+static const char millisFmt[]              = "%03" PRIu16;
+// static const char minutesSecondsBg[] = "88:88.";
+// static const char millisBg[] = "888";
+
+//==============================================================================
+// Variables
+//==============================================================================
+
+swadgeMode_t timerMode = {
+    .modeName                 = timerName,
+    .wifiMode                 = NO_WIFI,
+    .overrideUsb              = false,
+    .usesAccelerometer        = true,
+    .usesThermometer          = false,
+    .overrideSelectBtn        = false,
+    .fnEnterMode              = timerEnterMode,
+    .fnExitMode               = timerExitMode,
+    .fnMainLoop               = timerMainLoop,
+    .fnAudioCallback          = NULL,
+    .fnBackgroundDrawCallback = NULL,
+    .fnEspNowRecvCb           = NULL,
+    .fnEspNowSendCb           = NULL,
+    .fnAdvancedUSB            = NULL,
+};
+
+static timerMode_t* timerData = NULL;
+
+//==============================================================================
+// Functions
+//==============================================================================
+
+static void timerEnterMode(void)
+{
+    timerData = calloc(1, sizeof(timerMode_t));
+
+    for (uint8_t i = 0; i < HOURGLASS_FRAMES; i++)
+    {
+        char wsgName[18];
+        snprintf(wsgName, sizeof(wsgName), hourglassFrameFmt, i);
+        loadWsg(wsgName, &timerData->hourglassFrames[i], false);
+    }
+
+    loadFont("ibm_vga8.font", &timerData->textFont, false);
+    loadFont("seven_segment.font", &timerData->numberFont, false);
+
+    // 100FPS? Sure?
+    setFrameRateUs(1000000 / 100);
+
+    // Default to 30s
+    timerData->countdownTime = 30 * 1000000;
```

Comment:
Without instructions, I thought it would only count down from 30. Then I pressed more buttons!